### PR TITLE
WIP: Implement Role mechanic

### PR DESCRIPTION
**Pull request by theelk801**
Alright, this is basically done but I need to write tests for it.

### CODE REVIEW

**Comment by theelk801** [comment]
@xenohedron @Susucre lemme know if you have thoughts

File: Mage/src/main/java/mage/constants/RoleType.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import mage.game.permanent.token.Token;
+import mage.game.permanent.token.YoungHeroRoleToken;
+
+import java.util.UUID;
+import java.util.function.Supplier;
+
+/**
+ * @author TheElk801
+ */
+public enum RoleType {
```
Comment (Susucre) [inline Mage/src/main/java/mage/constants/RoleType.java:16]: I am conflicted on that.
I like the enum grouping with each instance linking to its token, don't get me wrong.
However I do not think this should be categorized/named as a Type.

File: Mage/src/main/java/mage/game/GameImpl.java [hunk 2]
```diff
@@ -2733,6 +2735,28 @@ protected boolean checkStateBasedActions() {
...
+                    .stream()
+                    .map(Map::values)
+                    .flatMap(Collection::stream)
+                    .filter(s -> s.size() > 1)
+                    .collect(Collectors.toList());
+            if (!rolesToHandle.isEmpty()) {
+                for (Set<Permanent> roleSet : rolesToHandle) {
+                    int newest = roleSet.stream().mapToInt(Permanent::getCreateOrder).max().orElse(-1);
+                    // using the world enchantment rule for ties until otherwise corrected
+                    if (roleSet.stream().mapToInt(Permanent::getCreateOrder).filter(i -> i == newest).count() < 2) {
```
Comment (Susucre) [inline Mage/src/main/java/mage/game/GameImpl.java:2749]: I may be missing something, but what is that condition is checking exactly? Can there be more than 2 roles with the same createOrder?

File: Mage/src/main/java/mage/game/permanent/token/YoungHeroRoleToken.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import mage.filter.FilterPermanent;
+import mage.filter.common.FilterCreaturePermanent;
+import mage.filter.predicate.mageobject.ToughnessPredicate;
+import mage.target.TargetPermanent;
+import mage.target.common.TargetCreaturePermanent;
+
+/**
+ * @author TheElk801
+ */
+public final class YoungHeroRoleToken extends TokenImpl {
```
Comment (Susucre) [inline Mage/src/main/java/mage/game/permanent/token/YoungHeroRoleToken.java:24]: I wonder if the reminder text are always the same for each kind of role.
At which point we could add those either in the Token definition (with an interface for a getter), or in the enum grouping them.
Probably a consideration to have with the full set spoiled.

**Review by Susucre — COMMENTED** [review]
Seems great to me. With the current info with have on the mechanic.

Edit: ok a few bugs to fix. I am apparently better as a live tester than at as a code reviewer.

File: Mage/src/main/java/mage/game/GameImpl.java (hunk 2 above)
Comment (theelk801) [reply Mage/src/main/java/mage/game/GameImpl.java:2749]: yeah I'm pretty sure it doesn't do anything and maybe it should just be removed

File: Mage/src/main/java/mage/constants/RoleType.java (hunk 1 above)
Comment (Susucre) [reply Mage/src/main/java/mage/constants/RoleType.java:16]: Nevermind, ignore me, this is exactly like TokenType

File: Mage/src/main/java/mage/game/permanent/token/YoungHeroRoleToken.java (hunk 3 above)
Comment (theelk801) [reply Mage/src/main/java/mage/game/permanent/token/YoungHeroRoleToken.java:24]: we might be able to include the reminder text into the enum but if it were up to me we wouldn't have any reminder text in generated rules text at all anywhere

File: Mage.Sets/src/mage/cards/c/CursedCourtier.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        this.subtype.add(SubType.HUMAN);
+        this.subtype.add(SubType.NOBLE);
+        this.power = new MageInt(3);
+        this.toughness = new MageInt(3);
+
+        // Lifelink
+        this.addAbility(LifelinkAbility.getInstance());
+
+        // When Cursed Courtier enters the battlefield, create a Cursed Role token attached to it.
+        this.addAbility(new EntersBattlefieldTriggeredAbility(new CreateRoleAttachedSourceEffect(RoleType.CURSED)));
```
Comment (Susucre) [inline Mage.Sets/src/mage/cards/c/CursedCourtier.java:32]: Just tried it, does not work as expected, you can select a different creature to attach the Role to it. Will try to reread your code and find the problem.

File: Mage/src/main/java/mage/game/GameImpl.java (hunk 2 above)
Comment (Susucre) [reply Mage/src/main/java/mage/game/GameImpl.java:2749]: Testing it, the role being sacrificed seems to be the newest one on all my attempts.
The one needing to be removed is the newest, the rest should be moved to the graveyard.

**Comment by theelk801** [comment]
ok, so this should be done now except for a bug that predates it regarding [[Estrid, the Masked]], [[Smoke Spirits' Aid]], and [[Stangg, Echo Warrior]] where aura tokens created attached to permanents are requiring that they be created attached but I think that can be a separate issue

File: Mage/src/main/java/mage/game/GameImpl.java (hunk 2 above)
Comment (theelk801) [reply Mage/src/main/java/mage/game/GameImpl.java]: this is fixed

File: Mage.Sets/src/mage/cards/c/CursedCourtier.java (hunk 4 above)
Comment (theelk801) [reply Mage.Sets/src/mage/cards/c/CursedCourtier.java:32]: this seems to be a preexisting bug, see #10035

Comment (Susucre) [reply Mage.Sets/src/mage/cards/c/CursedCourtier.java:32]: When I tried it earlier, it ended up with the aura attached to something it should not have (e.g. another creature with Cursed Courtier)
Does the change source.getFirstTarget() -> permanent.getId() fixed that to be the expected attached permanent?

Comment (theelk801) [reply Mage.Sets/src/mage/cards/c/CursedCourtier.java:32]: I believe so, yes

**Comment by Susucre** [comment]
Btw Travis will fail due to mtgjson having updated some of the verify data. Got that cleaned up there #10827

**Comment by theelk801** [comment]
if this passes tests I'm gonna merge it and then fix #10830 in a separate PR since that's a separate issue that already existed

**Pull request merged**